File: numerouno/numerouno/page/invoice_print_tracker/invoice_print_tracker.js
frappe.pages["invoice-print-tracker"].on_page_load = function (wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __("Sales Invoice Tracking"),
		single_column: true,
	});

	new SalesInvoiceTrackingPage(page);
};

class SalesInvoiceTrackingPage {
	constructor(page) {
		this.page = page;
		this.rows = [];
		this.print_format = "Sales Invoice NUTC";
		this.make_filters();
		this.make_layout();
		this.bind_actions();
		this.refresh();
	}

	make_filters() {
		this.view_filter = this.page.add_field({
			fieldname: "view",
			label: __("Show"),
			fieldtype: "Select",
			options: [
				{ value: "needs_print", label: __("Needs Print") },
				{ value: "done", label: __("Printed / Done") },
				{ value: "all", label: __("All Invoices") },
			],
			default: "needs_print",
			change: () => this.refresh(),
		});
		this.customer_filter = this.page.add_field({
			fieldname: "customer",
			label: __("Customer"),
			fieldtype: "Link",
			options: "Customer",
			change: () => this.refresh(),
		});
		this.from_date_filter = this.page.add_field({
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
			default: frappe.datetime.add_months(frappe.datetime.get_today(), -1),
			change: () => this.refresh(),
		});
		this.to_date_filter = this.page.add_field({
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
			default: frappe.datetime.get_today(),
			change: () => this.refresh(),
		});
		this.search_filter = this.page.add_field({
			fieldname: "search",
			label: __("Search"),
			fieldtype: "Data",
			change: () => this.refresh(),
		});
	}

	make_layout() {
		this.$root = $(`
			<div class="sit-page">
				
				<div class="sit-stats row text-muted small mb-3"></div>
				<div class="sit-table-wrap"></div>
			</div>
		`);
		this.page.main.append(this.$root);
		this.$stats = this.$root.find(".sit-stats");
		this.$table_wrap = this.$root.find(".sit-table-wrap");
	}

	bind_actions() {
		this.page.set_primary_action(__("Save"), () => this.save(), "save");
		this.page.add_inner_button(__("Refresh"), () => this.refresh(), "refresh");
		this.page.add_inner_button(__("Print Selected"), () => this.print_selected(), "print");
	}

	get_filters() {
		return {
			view: this.view_filter.get_value() || "needs_print",
			customer: this.customer_filter.get_value(),
			from_date: this.from_date_filter.get_value(),
			to_date: this.to_date_filter.get_value(),
			search: this.search_filter.get_value(),
		};
	}

	refresh() {
		frappe.call({
			method: "numerouno.numerouno.page.invoice_print_tracker.invoice_print_tracker.get_invoices",
			args: { filters: this.get_filters() },
			freeze: true,
			callback: (r) => {
				if (r.exc) return;
				const msg = r.message || {};
				this.rows = (msg.rows || []).map((row) => ({ ...row, _selected: false }));
				this.print_format = msg.print_format || this.print_format;
				this.render_stats(msg);
				this.render_table();
			},
		});
	}

	render_stats(msg) {
		const view = this.get_filters().view;
		let hint = __("Showing {0} invoice(s)", [msg.count || 0]);
		if (view === "needs_print") {
			hint += ` — ${__("invoices waiting to be printed")}`;
		}
		this.$stats.html(`<div class="col-12">${hint}</div>`);
	}

	render_table() {
		if (!this.rows.length) {
			this.$table_wrap.html(
				`<div class="text-muted p-4 text-center">${__("No invoices match this filter.")}</div>`
			);
			return;
		}

		const header = `
			<table class="table table-bordered table-hover sit-table">
				<thead>
					<tr>
						<th><input type="checkbox" class="sit-select-all" title="${__("Select")}"></th>
						<th>${__("Invoice")}</th>
						<th>${__("Customer")}</th>
						<th>${__("Date")}</th>
						<th>${__("Grand Total")}</th>
						<th>${__("Print Status")}</th>
						<th>${__("Email Sent")}</th>
						<th>${__("Original Sent")}</th>
						<th>${__("Remarks")}</th>
						<th></th>
					</tr>
				</thead>
				<tbody></tbody>
			</table>`;
		this.$table_wrap.html(header);
		const $tbody = this.$table_wrap.find("tbody");

		this.rows.forEach((row, idx) => {
			const status = row.printed
				? `<span class="indicator-pill green">${__("Printed")}</span>`
				: `<span class="indicator-pill orange">${__("Not printed")}</span>`;
			const printed_meta = row.printed_on
				? `<div class="text-muted small">${frappe.datetime.str_to_user(row.printed_on)}</div>`
				: "";
			$tbody.append(`
				<tr data-idx="${idx}">
					<td><input type="checkbox" class="sit-row-select" data-idx="${idx}"></td>
					<td><a href="/app/sales-invoice/${encodeURIComponent(row.invoice_number)}">${frappe.utils.escape_html(
						row.invoice_number
					)}</a></td>
					<td>${frappe.utils.escape_html(row.customer_name || row.customer || "")}</td>
					<td>${frappe.datetime.str_to_user(row.invoice_date)}</td>
					<td>${frappe.format(row.grand_total, { fieldtype: "Currency", options: row.currency })}</td>
					<td>${status}${printed_meta}</td>
					<td class="text-center"><input type="checkbox" class="sit-email" data-idx="${idx}" ${
						row.email_sent ? "checked" : ""
					}></td>
					<td class="text-center"><input type="checkbox" class="sit-original" data-idx="${idx}" ${
						row.original_sent ? "checked" : ""
					}></td>
					<td><input type="text" class="form-control input-sm sit-remarks" data-idx="${idx}" value="${frappe.utils.escape_html(
						row.remarks || ""
					)}"></td>
					<td><button class="btn btn-default btn-xs sit-print-one" data-invoice="${frappe.utils.escape_html(
						row.invoice_number
					)}">${__("Print")}</button></td>
				</tr>
			`);
		});

		this.$table_wrap.find(".sit-select-all").on("change", (e) => {
			const checked = e.target.checked;
			this.$table_wrap.find(".sit-row-select").prop("checked", checked);
		});

		this.$table_wrap.find(".sit-print-one").on("click", (e) => {
			const invoice = $(e.currentTarget).data("invoice");
			this.print_invoices([invoice], true);
		});
	}

	collect_rows_for_save() {
		return this.rows.map((row, idx) => {
			const $tr = this.$table_wrap.find(`tr[data-idx="${idx}"]`);
			if (!$tr.length) return null;
			return {
				invoice_number: row.invoice_number,
				email_sent: $tr.find(".sit-email").is(":checked") ? 1 : 0,
				original_sent: $tr.find(".sit-original").is(":checked") ? 1 : 0,
				remarks: $tr.find(".sit-remarks").val(),
				printed: row.printed ? 1 : 0,
			};
		}).filter(Boolean);
	}

	get_selected_invoices() {
		const selected = [];
		this.$table_wrap.find(".sit-row-select:checked").each((_, el) => {
			const idx = $(el).data("idx");
			if (this.rows[idx]) selected.push(this.rows[idx].invoice_number);
		});
		return selected;
	}

	save() {
		const payload = this.collect_rows_for_save();
		frappe.call({
			method: "numerouno.numerouno.page.invoice_print_tracker.invoice_print_tracker.save_tracking_rows",
			args: { rows: payload },
			freeze: true,
			callback: (r) => {
				if (r.exc) return;
				frappe.show_alert({
					message: __("Saved {0} row(s)", [(r.message || {}).updated || 0]),
					indicator: "green",
				});
				this.refresh();
			},
		});
	}

	print_selected() {
		const invoices = this.get_selected_invoices();
		if (!invoices.length) {
			frappe.msgprint(__("Select at least one invoice to print."));
			return;
		}
		this.print_invoices(invoices, true);
	}

	print_invoices(invoices, mark_printed) {
		invoices.forEach((name) => {
			const url = frappe.urllib.get_full_url(
				`/printview?doctype=${encodeURIComponent("Sales Invoice")}&name=${encodeURIComponent(
					name
				)}&format=${encodeURIComponent(this.print_format)}&trigger_print=1`
			);
			window.open(url, "_blank");
		});

		if (!mark_printed) return;

		frappe.call({
			method: "numerouno.numerouno.page.invoice_print_tracker.invoice_print_tracker.mark_printed",
			args: { invoice_numbers: invoices, open_print: 0 },
			callback: (r) => {
				if (r.exc) return;
				frappe.show_alert({
					message: __("Marked {0} invoice(s) as printed", [((r.message || {}).marked || []).length]),
					indicator: "green",
				});
				this.refresh();
			},
		});
	}
}

File: numerouno/numerouno/page/invoice_print_tracker/invoice_print_tracker.py
import json

import frappe
from frappe import _
from frappe.utils import getdate, now_datetime

PRINT_FORMAT = "Sales Invoice NUTC"


def _filters_sql(filters):
	conditions = ["si.docstatus = 1"]
	values = {}

	if filters.get("customer"):
		conditions.append("si.customer = %(customer)s")
		values["customer"] = filters.get("customer")

	if filters.get("from_date"):
		conditions.append("si.posting_date >= %(from_date)s")
		values["from_date"] = getdate(filters.get("from_date"))

	if filters.get("to_date"):
		conditions.append("si.posting_date <= %(to_date)s")
		values["to_date"] = getdate(filters.get("to_date"))

	if filters.get("search"):
		conditions.append("(si.name LIKE %(search)s OR si.customer_name LIKE %(search)s)")
		values["search"] = f"%{filters.get('search').strip()}%"

	view = (filters.get("view") or "needs_print").strip()
	if view == "needs_print":
		conditions.append("(sit.name IS NULL OR IFNULL(sit.printed, 0) = 0)")
	elif view == "done":
		conditions.append("IFNULL(sit.printed, 0) = 1")

	return " AND ".join(conditions), values


@frappe.whitelist()
def get_invoices(filters=None):
	filters = frappe._dict(json.loads(filters) if isinstance(filters, str) else (filters or {}))
	where, values = _filters_sql(filters)

	rows = frappe.db.sql(
		f"""
		SELECT
			si.name AS invoice_number,
			si.customer,
			si.customer_name,
			si.posting_date AS invoice_date,
			si.grand_total,
			si.currency,
			sit.name AS tracking_name,
			IFNULL(sit.printed, 0) AS printed,
			sit.printed_on,
			sit.printed_by,
			IFNULL(sit.email_sent, 0) AS email_sent,
			IFNULL(sit.original_sent, 0) AS original_sent,
			sit.remarks
		FROM `tabSales Invoice` si
		LEFT JOIN `tabSales Invoice Tracking` sit
			ON sit.invoice_number = si.name
			AND sit.docstatus < 2
		WHERE {where}
		ORDER BY si.posting_date DESC, si.name DESC
		LIMIT 500
		""",
		values,
		as_dict=True,
	)

	needs = sum(1 for r in rows if not r.printed)
	return {
		"rows": rows,
		"count": len(rows),
		"needs_print_in_result": needs,
		"print_format": PRINT_FORMAT,
	}


def _get_or_create_tracking(invoice_number: str):
	name = frappe.db.get_value(
		"Sales Invoice Tracking",
		{"invoice_number": invoice_number, "docstatus": ("<", 2)},
		"name",
	)
	if name:
		return frappe.get_doc("Sales Invoice Tracking", name)

	doc = frappe.new_doc("Sales Invoice Tracking")
	doc.invoice_number = invoice_number
	doc.insert(ignore_permissions=True)
	return doc


@frappe.whitelist()
def save_tracking_rows(rows=None):
	"""Bulk save email/original flags from the workbench."""
	rows = json.loads(rows) if isinstance(rows, str) else (rows or [])
	updated = 0
	for row in rows:
		invoice = (row.get("invoice_number") or "").strip()
		if not invoice:
			continue
		doc = _get_or_create_tracking(invoice)
		doc.email_sent = 1 if row.get("email_sent") else 0
		doc.original_sent = 1 if row.get("original_sent") else 0
		if row.get("remarks") is not None:
			doc.remarks = row.get("remarks") or ""
		if row.get("printed"):
			doc.printed = 1
			if not doc.printed_on:
				doc.printed_on = now_datetime()
				doc.printed_by = frappe.session.user
		doc.save(ignore_permissions=True)
		updated += 1
	frappe.db.commit()
	return {"updated": updated}


@frappe.whitelist()
def mark_printed(invoice_numbers=None, open_print=0):
	invoices = json.loads(invoice_numbers) if isinstance(invoice_numbers, str) else invoice_numbers
	if not invoices:
		frappe.throw(_("Select at least one invoice."))

	marked = []
	for invoice in invoices:
		invoice = (invoice or "").strip()
		if not invoice:
			continue
		doc = _get_or_create_tracking(invoice)
		doc.printed = 1
		doc.printed_on = now_datetime()
		doc.printed_by = frappe.session.user
		doc.save(ignore_permissions=True)
		marked.append(invoice)

	frappe.db.commit()
	return {
		"marked": marked,
		"print_format": PRINT_FORMAT,
		"open_print": int(open_print or 0),
	}
